modules/imgproc/src/hough.cpp

```diff
@@ -882,6 +881,700 @@ void cv::HoughLinesP(InputArray _image, OutputArray _lines,
+
+                    for(; x <= numCols - 32; x += 32) {
+                        v_uint8x16 v_edge1 = v_load(edgeData + x);
+                        v_uint8x16 v_edge2 = v_load(edgeData + x + 16);
+
+                        v_uint8x16 v_cmp1 = (v_edge1 == v_zero);
+                        v_uint8x16 v_cmp2 = (v_edge2 == v_zero);
+
+                        unsigned int mask1 = v_signmask(v_cmp1);
+                        unsigned int mask2 = v_signmask(v_cmp2);
+
+                        mask1 ^= 0x0000ffff;
+                        mask2 ^= 0x0000ffff;
+
+                        if(mask1)
+                        {
+                            x += trailingZeros32(mask1);
+                            goto _next_step;
+                        }
+
+                        if(mask2)
+                        {
+                            x += trailingZeros32(mask2 << 16);
+                            goto _next_step;
+                        }
+                    }
+                }
+#endif
+                for(; x < numCols && !edgeData[x]; ++x)
+                    ;
+
+                if(x == numCols)
+                    continue;
+#if CV_SIMD128
+_next_step:
+#endif
+                float vx, vy;
+                int sx, sy, x0, y0, x1, y1;
+
+                vx = dxData[x];
+                vy = dyData[x];
+
+                if(vx == 0 && vy == 0)
+                    continue;
+
+                float mag = std::sqrt(vx*vx+vy*vy);
+
+                if(mag < 1.0f)
+                    continue;
+
+                Point pt = Point(x % edges.cols, y + x / edges.cols);
+                nzLocal.push_back(pt);
+
+                sx = cvRound((vx * idp) * 1024 / mag);
+                sy = cvRound((vy * idp) * 1024 / mag);
+
+                x0 = cvRound((pt.x * idp) * 1024);
+                y0 = cvRound((pt.y * idp) * 1024);
+
+                // Step from min_radius to max_radius in both directions of the gradient
+                for(int k1 = 0; k1 < 2; k1++ )
+                {
+                    x1 = x0 + minRadius * sx;
+                    y1 = y0 + minRadius * sy;
+
+                    for(int r = minRadius; r <= maxRadius; x1 += sx, y1 += sy, r++ )
+                    {
+                        int x2 = x1 >> 10, y2 = y1 >> 10;
+                        if( (unsigned)x2 >= (unsigned)acols ||
+                                (unsigned)y2 >= (unsigned)arows )
+                            break;
+
+                        adataLocal[y2*astep + x2]++;
+                    }
+
+                    sx = -sx; sy = -sy;
+                }
+            }
+        }
+
+        if (singleThread)
+        {
+            accum = accumLocal;
+            nz = nzLocal;
+        }
+        else
+        {
+            AutoLock lock(_lock);
+            if(accum.empty())
+            {
+                accum = accumLocal;
+            }
+            else
+            {
+                add(accum, accumLocal, accum);
```

Comment:
Processing cv::add() under locked mutex is not a good thing anyway.
Consider to use thread local containers with reduction.